### PR TITLE
Fix for creatures using a MC spell on a single player

### LINKED ISSUE #546: Salia in Felwood casts a mind control spell, resets fight when solo
--

Salia in Felwood casts Cause Insanity, a form of mind control, even if only one character is attacking. This resets the fight. Since the spell is cast almost immediately when entering combat, it's nearly impossible to complete the fight solo. Cause Insanity should only be cast when two or more characters are engaged.

Tested on priest, can't confirm how player pets affect things.

allakhazam link for patch 1.12.1 here:
[https://web.archive.org/web/20060614164705/http://wow.allakhazam.com:80/db/mob.html?wmob=9860](https://web.archive.org/web/20060614164705/http://wow.allakhazam.com:80/db/mob.html?wmob=9860)

### PR DESCRIPTION
**Pull request by SpaceVector**
Fixes #546
**EDIT2  : It appears that hunters/warlocks and anything with a pet could get mind controlled back then in vanilla, so there's no mechanic involving the number of player threats. Only the number of threats matters. Also, Dr. Weavil could mind control a single player (meaning a threat list size of 1) in vanilla although his spell applies SPELL_AURA_MOD_CHARM (I checked the DBC file and tested in-game also), so solution 2 isn't blizzlike and we must use the new flag added to CastFlags to cherry-pick which spells we want to prevent from being cast on a single player, which this PR already took care of and is ready for merge.**

~~EDIT : This PR actually makes the fight against Salia easier for warlock/hunters since they would never get mind controlled during the fight if they're soloing with their pet. We need confirmation that warlocks/hunters, or anyone with a pet, wouldn't get mind controlled in retail if the given creature only has the player and its pet in its threat list.~~

This PR adds support for creatures that can mind control a single player, and more broadly, it adds support for creature spells that shouldn't be cast on a single player.

What happened was that the player wasn't hostile anymore if he got mind controlled and was removed from the threat list of all creatures that had the player on their threat list, thus making all creatures evade ,which was a game breaking bug in some places (like for Salia, which I fixed here) as you could enter a mind control loop.

A Blizzlike behavior is that the creature won't cast any MC spell on a player that is alone.

To fix this, we could either : 
1) Add a new flag to CastFlags that we check whenever a creature will cast a spell. Then we can set this flag for any creature spell in its AI and fix this issue just by editing the database.

2) Prevent casting if the spell applies the following auras : SPELL_AURA_MOD_POSSESS or SPELL_AURA_MOD_CHARM

Solution 2 seems more generic but this PR goes for the flag solution so we can cherry-pick which creature spells we want to prevent being cast on a single player, since there was at least one exception back in retail : 

> There was at least one NPC who could mind control a solo player in Vanilla.
http://wowwiki.wikia.com/wiki/Dr._Weavil

>"Note that he cannot be soloed by any class without a combat pet, and the only classes able to get away from an endless loop of mind control are: Shaman, Hunter, Warlock, Paladin, and Mage, as Weavil will reset and return to full health when all opponents are dead or mind controlled. (See Tips for methods to escape). "

>Of course it is impossible to solo him with a level 60 player but still technically this NPC never has any mechanic to avoid MC a solo player.

I suspect Blizzard just forgot to add the right auras for Dr. Weavil's spell or the flag for it, depending on their implementation. Considering we want a blizzlike behavior, solution 1 is indeed better.

Since I have a modest knowledge of vanilla, I also went for the flag solution in case there are other kinds of spell (not necessarily a MC spell) that shouldn't be cast on a single player, but I don't think there is any other.

In this PR, I only fixed Salia. But there might be other creatures that share this problem. We can now just update their spell CastFlags in their AI to fix it.

### CODE REVIEW

File: src/game/Threat/ThreatManager.cpp [hunk 1]
```diff
@@ -557,3 +557,17 @@ void ThreatManager::setCurrentVictimIfCan(Unit * pVictim)
...
     }
 }
+
+uint8 ThreatManager::numberOfPlayerThreats() const
+{
+    uint8 player_amount = 0;
+    ThreatList const& threatList = getThreatList();
+    for (ThreatList::const_iterator itr = threatList.begin(); itr != threatList.end(); itr++)
+    {
+        if (iOwner->GetMap()->GetUnit((*itr)->getUnitGuid())->IsPlayer())
```
Comment (thepparker) [inline src/game/Threat/ThreatManager.cpp]: You don't need to do a map lookup here, you can simply check the high flag on the GUID by using (*itr)->getUnitGuid().IsPlayer()

Comment (SpaceVector) [reply src/game/Threat/ThreatManager.cpp]: Thanks, didn't notice it !

**Comment by Fiskehatt** [comment]
>  EDIT : This PR actually makes the fight against Salia easier for warlock/hunters since they would never get mind controlled during the fight if they're soloing with their pet. We need confirmation that warlocks/hunters, or anyone with a pet, wouldn't get mind controlled in retail if the given creature only has the player and its pet in its threat list.

Some vanilla [comments ](https://web.archive.org/web/20070406190022/http://www.thottbot.com:80/q5242)claim that you could get MC'd with a pet:

> No credit for kills
> Score 1.1     Vote: [-] [+] by dyl666, 1.5 years ago
> This really irked me the first two times I tried it - basically I was never getting credit for Killing Salia.
> 
> The second time I tried it I watched what was going on - and basically, when she casts her 'temporary insanity' on you (which a couple ppl have mentioned - makes you do random stuff; I started healing my pet at one point which was actually quite helpful!!)  you LOSE your 'tag' on her - i.e. she goes grey (as if someone else had been killing her.) When she did it on my I ended up attacking my own pet - not sure if this is the only circumstance she untags.
> 
> Anyway, I killed the other 2, dismissed my pet, and solo'd her, and got full credit for the kill - quest completed.
> 
> Hope this helps someone - it really pissed me off the first time(s) it happened. 

> Watch out..
> Score 1.4     Vote: [-] [+] by Warkha, 1.1 years ago
> One of the Succubi has the ability to mind control. She MC'd me towards the end of the fight as my pet killed her meaning i didnt get credit for the kill. She did this twice. Third Time i stood alot further back and it worked. This would really be only problem for a Lock/Hunter on this quest i feel. 

> But the succies were the ones that caused me the hassle. As i killed the orc, the two succies came for me, one of them i killed with a few melee hits, (had little health left from my shots), then the second one, had pretty much full health, but sent me into a mind control curse, where i found myself attacking my puddy cat. 

You could possibly also get MC'd when soloing:

> 
> Thanks a bunch. I did exactly that and here's mage advice: sheep Fel'dan as he's the only humanoid. Wail on Moora as she's the healer. Frost nove the two succubi when you're doing this. **Salia has this mind control thing and will bring you back over to Fel'dan onc eMoora dies.** Begin wailing on him. Put up a shield. Once he dies, frost nova Salia and bandage. Then take her down.
> 
> As I did this as a lvl 59 fire mage, you may end up using potions, or kill one, die, then rez and kill the last. I shouldn't think it'd be too hard a quest, just keep a watch on your health bar. 

This comment is from TBC though.

Regarding Dr. Weavil, this [video ](https://www.youtube.com/watch?v=SktazVwEWvw) (Wotlk) claims that he always MC's the person with the highest threat:

> Usually impossible to solo because of his mechanics. He will MC top aggro, meaning the player if they are alone. You can get around this by pet tanking, as he will not MC a pet. 

I hope this evidence is useful in figuring out how the mechanic works 😅

**Comment by SpaceVector** [comment]
Thanks for the info. I guess only the threat list size matters then. There's no need to count any player.

We should check if Dr. Weavil's MC ability applies any aura (I'm not really used to browse dbc files but that should be easy enough) specified in Solution #2. Guess that's just an oversight by Blizzard back then and there's no need for any flag.

**Comment by brotalnia** [comment]
I've added a check for this in CreatureAI::CanCastSpell as part of my creature_spells pull request. It's better to have it there as it should work this way by default. For special cases where you want the spell to be cast anyway you can just use the force cast flag. Also it was already possible to make it so the spell is not casted with just 1 target by using TARGET_T_HOSTILE_RANDOM_NOT_TOP.

**Comment by brotalnia** [comment]
Closing this as it's been implemented in a different way.

**Pull request closed without merging**